src/test-helpers/e2e-helpers/clickWithOffset.ts

```diff
@@ -0,0 +1,42 @@
+import { JSHandle, Page } from 'puppeteer'
+
+interface Options {
+  // Where in the horizontal line (inside) of the target node should the mouse be clicked
+  horizontalClickLine?: 'left' | 'right',
+  // Specify specific node on editable to click on. Overrides horizontalClickLine
+  offset?: number,
+}
+
+/**
+ * Click the given node with offset.
+ */
+const clickEditable = async (page: Page, nodeHandle: JSHandle, { horizontalClickLine = 'left', offset }: Options) => {
+
+  const boundingBox = await nodeHandle.asElement()?.boundingBox()
+
+  if (!boundingBox) throw new Error('Bouding box of editable not found.')
+
+  /** Get cordinates for specific text node if the given node has text child. */
+  const offsetCoordinates = () => page.evaluate((node: HTMLElement, offset: number) => {
+    const textNode = node.firstChild
+    if (!textNode || textNode.nodeName !== '#text') return
+    const range = document.createRange()
+    range.setStart(textNode, offset)
+    const { right, top, height } = range.getBoundingClientRect()
+    return {
+      x: right,
+      y: top + (height / 2)
+    }
+  }, nodeHandle, offset || 0)
+
+  const coordinate = !offset ? {
+    x: boundingBox.x + (horizontalClickLine === 'left' ? 0 : horizontalClickLine === 'right' ? boundingBox.width : boundingBox.width / 2),
```

Comment:
It looks like "left" still clicks on the left edge of the element instead of to the left of the element (outside). Likewise for "right".

Reply:
I added a `HORIZONTAL_OFFSET` in 2135e4ced but the "right" test now fails. Can you investigate?

Reply:
@raineorshine I am a little confused here. By clicking right of the thought. Do you mean to the right(outside) of the `Editable`? If that is the case it activates `cursorUp` instead of caret being set at the end of the editable ?

Reply:
I don’t think it should activate cursorUp when clicking to the right of a thought. Does the Editable itself go all the way to the right edge? I thought it might have been a container element or ThoughtAnnotation. I’m not at my computer so I can’t test. 

On mobile the caret correctly goes to the end of the thought if the cursor was previously on a sibling. If the cursor is not on a sibling, it incorrectly puts the cursor at the beginning.

Reply:
It's working on mobile. However, I still cannot get consistent results. It's putting caret at the beginning sometimes. Should the caret always be positioned at the end of the thought if clicked anywhere right (outside) of the editable? Or does it depend on where the previous cursor was?

Reply:
https://user-images.githubusercontent.com/25879100/114599822-55f30100-9cb3-11eb-8b5d-81ba045c4039.mov

Reply:
I am using a simulator here. I have highlighted the editable. I am getting inconsistent caret placement on clicking on the right side of the editable. Am I missing something here ?

Reply:
> It's working on mobile.

On iOS, the caret goes to the beginning of a thought instead of the end when the cursor is on a non-sibling before the tap. I will create a separate issue for this, though I remember there being some complication with it when I tried to solve it before.

> Should the caret always be positioned at the end of the thought if clicked anywhere right (outside) of the editable? 

If user clicks slightly to the right of the text, it should definitely place the caret at the end of the editable. For the dead space way to the right, I'm not sure, but I've never clicked there and expected something different to happen, so I say leave it for now. I added `.skip` to the test for clicking to the right.

> Or does it depend on where the previous cursor was?

The caret placement should never depend on where the cursor was previously (except for determining edit mode).

> I am getting inconsistent caret placement on clicking on the right side of the editable. Am I missing something here ?

I can't explain the inconsistent behavior, but I think it's okay to ignore it since that has not caused any UX issues that I have noticed.